### PR TITLE
[Gen2] Add cellular registration intervention commands to Gen2

### PR DESCRIPTION
**Pull request by keeramis**
### Problem

Add cellular AT intervention commands to Gen2, to improve/speed up registration process. This is particularly helpful in scenarios where the radio struggles to connect to an invalid operator.

### Solution

1. Added `CellularRegistrationStatusMdm` similar to the one in Gen3 to run intervention commands
2. Replaced the `csd, psd, and eps` members of `NetStatus` with `CellularRegistrationStatusMdm`'s status methods for consistency.

### Steps to Test

Run any tinker-like app on Gen3 which enables connecting to the network. Try removing the antenna or using an invalid SIM, even better if there are known network issues in a given area to check how the the intervention commands run and help register to a network.

### Example App

```c
void setup() {
}

void loop() {
}
```

### References

- [ch69276](https://app.clubhouse.io/particle/story/69276/port-gen-3-cellular-registration-intervention-logic-to-gen2)

---

### Completeness

- [x] User is totes amazing for contributing!
- [x] Contributor has signed CLA ([Info here](https://github.com/spark/firmware/blob/develop/CONTRIBUTING.md))
- [x] Problem and Solution clearly stated
- [ ] Run unit/integration/application tests on device
- [ ] Added documentation
- [ ] Added to CHANGELOG.md after merging (add links to docs and issues)

### CODE REVIEW

File: hal/src/electron/modem/mdm_hal.h [hunk 1]
```diff
@@ -30,6 +30,105 @@
 #include "system_tick_hal.h"
 #include "cellular_enums_hal.h"
 #include "mdm_debug.h"
+#include "timer_hal.h"
+
+class CellularRegistrationStatusMdm {
```
Comment (avtolstoy) [inline hal/src/electron/modem/mdm_hal.h:35]: This is one to one to `CellularRegistrationStatus` in `cellular_ncp_client.h`. Let's just move the class into a new common header under `shared` to be used universally across Gen 2 and Gen 3.

File: hal/src/electron/modem/mdm_hal.cpp [hunk 2]
```diff
@@ -1554,6 +1569,105 @@ int MDMParser::_cbURAT(int type, const char *buf, int len, bool *matched_default
...
+            if (psd_.status() == CellularRegistrationStatusMdm::NOT_REGISTERING) {
+                MDM_INFO("Sticky not registering PSD state for %lu s, force GPRS attach", psd_.duration() / 1000);
+                psd_.reset();
+                _registrationInterventions++;
+                sendFormated("AT+CGACT?\r\n");
+                if (WAIT == waitFinalResp(nullptr, nullptr, CGACT_TIMEOUT)) {
+                    return false;
+                }
+                sendFormated("AT+CGACT=1\r\n");
+                if (RESP_OK != waitFinalResp(nullptr, nullptr, CGACT_TIMEOUT)) {
```
[the remark this answers is not in the document]
Comment (keeramis) [reply hal/src/electron/modem/mdm_hal.cpp]: Used look up using the resp code itself. Please let me know if `CHECK_TIMEOUT` is better. Will try something like this perhaps?

```
                sendFormated("AT+CGATT=1\r\n");
                if (RESP_OK != CHECK_TIMEOUT(waitFinalResp(nullptr, nullptr, CGATT_TIMEOUT))) {
                    csd_.reset();
                    psd_.reset();
                    if (!_checkModem()) {
                        return false;
                    }
                    MDM_INFO("GPRS attach failed, try PLMN reselection");
                    sendFormated("AT+COPS=0,2\r\n");
                    if (WAIT == waitFinalResp(nullptr, nullptr, COPS_TIMEOUT)) {
                        return false;
                    }
                }
```

File: hal/src/electron/modem/mdm_hal.h (hunk 1 above)
Comment (keeramis) [reply hal/src/electron/modem/mdm_hal.h]: Good idea. 👍  Added.

**Review by avtolstoy — APPROVED** [review]

File: hal/src/electron/modem/mdm_hal.h [hunk 3]
```diff
@@ -30,6 +30,8 @@
 #include "system_tick_hal.h"
 #include "cellular_enums_hal.h"
 #include "mdm_debug.h"
+#include "timer_hal.h"
```
Comment (technobly) [inline hal/src/electron/modem/mdm_hal.h:33]: Nit: extra to add timer_hal.h?

**Review by technobly — APPROVED** [review]

File: hal/shared/cellular_enums_hal.h [hunk 4]
```diff
@@ -70,14 +70,7 @@ typedef struct {
...
 } DevStatus;
 //! Registration Status
-typedef enum {
-    REG_UNKNOWN = 0,
-    REG_DENIED  = 1,
-    REG_NONE    = 2,
-    REG_HOME    = 3,
-    REG_ROAMING = 4,
-    REG_NOTREG  = 5
-} Reg;
```
[the remark this answers is not in the document]
Comment (avtolstoy) [reply hal/shared/cellular_enums_hal.h:80]: A good question, however this enum specifically is not used anywhere in our APIs (neither in `cellular_hal` nor in `wiring`) so I would argue that we shouldn't need to keep it around for compatibility reasons.

[the remark this answers is not in the document]
Comment (technobly) [reply hal/shared/cellular_enums_hal.h:80]: I have used these before in user firmware which is what prompted the question, but it was a special build now that I think of it.  Probably unlikely these would be used.

**Pull request merged**